File: Components/DevExpressVCL/Demos/ExpressEditors Library/Delphi/RangeControlDemo/RangeControlDemoMain.pas
unit RangeControlDemoMain;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics,
  Controls, Forms, Dialogs, BaseForm, Menus, StdCtrls, Types,
  cxGraphics, cxControls, cxLookAndFeels, cxLookAndFeelPainters, dxRangeControl,
  cxContainer, cxEdit, dxBevel, cxGroupBox, Generics.Collections, Generics.Defaults,
  cxCheckBox, dxTypeHelpers, cxRadioGroup, cxTrackBar, cxLabel, cxTextEdit,
  cxMaskEdit, cxDropDownEdit, cxCheckComboBox, dxCore;

type
  TChartData = record
    X: Variant;
    Y: Integer;
  end;

  TdxRangeControlDemoForm = class(TfmBaseForm)
    cxGroupBox1: TcxGroupBox;
    rcNumericClient: TdxRangeControl;
    cxGroupBox2: TcxGroupBox;
    rcDateTimeClient: TdxRangeControl;
    cxGroupBox3: TcxGroupBox;
    rcDateTimeHeaderClient: TdxRangeControl;
    cxGroupBox4: TcxGroupBox;
    cxGroupBox6: TcxGroupBox;
    cxGroupBox7: TcxGroupBox;
    rgDateTimeClientContentType: TcxRadioGroup;
    rgNumberClientContentType: TcxRadioGroup;
    chbAutoFormatScaleCaptions: TcxCheckBox;
    tbNumericClient: TcxTrackBar;
    cxLabel1: TcxLabel;
    tbDateTimeClient: TcxTrackBar;
    cxLabel2: TcxLabel;
    cxGroupBox5: TcxGroupBox;
    Options1: TMenuItem;
    ShowRuler1: TMenuItem;
    ShowZoomscrollbar1: TMenuItem;
    Animation1: TMenuItem;
    cxCheckComboBox1: TcxCheckComboBox;
    cxLabel3: TcxLabel;
    procedure Animation1Click(Sender: TObject);
    procedure rcNumericClientDrawContent(ASender: TdxCustomRangeControl; ACanvas: TcxCanvas; AViewInfo: TdxRangeControlCustomClientViewInfo; var AHandled: Boolean);
    procedure FormDestroy(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure rcDateTimeClientDrawContent(ASender: TdxCustomRangeControl; ACanvas: TcxCanvas; AViewInfo: TdxRangeControlCustomClientViewInfo; var AHandled: Boolean);
    procedure rcDateTimeHeaderClientDrawContent(ASender: TdxCustomRangeControl; ACanvas: TcxCanvas; AViewInfo: TdxRangeControlCustomClientViewInfo; var AHandled: Boolean);
    procedure FormResize(Sender: TObject);
    procedure ShowRuler1Click(Sender: TObject);
    procedure ShowZoomscrollbar1Click(Sender: TObject);
    procedure tbDateTimeClientPropertiesChange(Sender: TObject);
    procedure rgNumberClientContentTypePropertiesEditValueChanged(
      Sender: TObject);
    procedure tbNumericClientPropertiesChange(Sender: TObject);
    procedure rgDateTimeClientContentTypePropertiesEditValueChanged(
      Sender: TObject);
    procedure chbAutoFormatScaleCaptionsPropertiesEditValueChanged(
      Sender: TObject);
    procedure cxCheckComboBox1PropertiesEditValueChanged(Sender: TObject);
  private
    FNumericClientData: TList<TChartData>;
    FDateTimeClientData2: TList<TChartData>;
    FDateTimeClientData1: TList<TChartData>;
    FDateTimeHeaderClientData1: TDictionary<TDateTime, Integer>;
    FDateTimeHeaderClientData2: TDictionary<TDateTime, Integer>;
  public
    procedure AddPolygonPoint(var APolygon: TPoints; X, Y: Integer);
    procedure InitializeChartData;
    function IsDateTimeClientContentLineMode: Boolean;
    function IsNumberClientContentLineMode: Boolean;
  end;

var
  dxRangeControlDemoForm: TdxRangeControlDemoForm;

implementation

uses
  Math, DateUtils, cxGeometry, dxGdiPlusClasses, dxCoreGraphics, DemoUtils;

{$R *.dfm}

procedure TdxRangeControlDemoForm.AddPolygonPoint(var APolygon: TPoints; X, Y: Integer);
begin
  SetLength(APolygon, Length(APolygon) + 1);
  APolygon[High(APolygon)].X := X;
  APolygon[High(APolygon)].Y := Y;
end;

procedure TdxRangeControlDemoForm.Animation1Click(Sender: TObject);
var
  AIsChecked: Boolean;
begin
  AIsChecked := dxDemoIsMenuChecked(Sender);
  rcNumericClient.Animation := AIsChecked;
  rcDateTimeClient.Animation := AIsChecked;
  rcDateTimeHeaderClient.Animation := AIsChecked;
end;

procedure TdxRangeControlDemoForm.chbAutoFormatScaleCaptionsPropertiesEditValueChanged(
  Sender: TObject);
begin
  (rcDateTimeHeaderClient.ClientProperties as TdxRangeControlDateTimeHeaderClientProperties).AutoFormatScaleCaptions := chbAutoFormatScaleCaptions.Checked;
end;

procedure TdxRangeControlDemoForm.cxCheckComboBox1PropertiesEditValueChanged(
  Sender: TObject);
var
  AProperties: TdxRangeControlDateTimeHeaderClientProperties;
  AScales: TdxRangeControlDateTimeScales;
  I: Integer;
begin
  AProperties := (rcDateTimeHeaderClient.ClientProperties as TdxRangeControlDateTimeHeaderClientProperties);
  AScales := AProperties.Scales;
  for I := 0 to cxCheckComboBox1.Properties.Items.Count - 1 do
    AScales.GetScale(TdxRangeControlDateTimeScaleUnit(I + Ord(rcduDay))).Visible := cxCheckComboBox1.States[I] = cbsChecked;
end;

procedure TdxRangeControlDemoForm.rcNumericClientDrawContent(ASender: TdxCustomRangeControl; ACanvas: TcxCanvas; AViewInfo: TdxRangeControlCustomClientViewInfo; var AHandled: Boolean);

  procedure DrawData(ADataSource: TList<TChartData>; const R: TRect; AColor: TdxAlphaColor);
  var
    AData: TChartData;
    APos, AValue: Integer;
    APolygon: TPoints;
  begin
    dxGPPaintCanvas.BeginPaint(ACanvas.Handle, AViewInfo.Bounds);
    dxGPPaintCanvas.SmoothingMode := smAntiAlias;
    try
      for AData in ADataSource do
        if InRange(AData.X, ASender.VisibleRangeMinValue - 1, ASender.VisibleRangeMaxValue + 1) then
        begin
          APos := ASender.GetPositionFromValue(AData.X);
          AValue := R.Bottom - AData.Y * R.Height div 100;
          if IsNumberClientContentLineMode then
            AddPolygonPoint(APolygon, APos, AValue)
          else
            dxGPPaintCanvas.Line(APos, R.Bottom, APos, AValue, AColor, 3);
        end;
      if IsNumberClientContentLineMode then
        dxGPPaintCanvas.Polyline(APolygon, AColor);
    finally
      dxGPPaintCanvas.EndPaint;
    end;
  end;

var
  R: TRect;
begin
  R := AViewInfo.Content.Bounds;
  DrawData(FNumericClientData, R, $FFEE8C4B);
end;

procedure TdxRangeControlDemoForm.rgDateTimeClientContentTypePropertiesEditValueChanged(
  Sender: TObject);
begin
  rcDateTimeClient.ContentChanged;
end;

procedure TdxRangeControlDemoForm.rgNumberClientContentTypePropertiesEditValueChanged(
  Sender: TObject);
begin
  rcNumericClient.ContentChanged;
end;

procedure TdxRangeControlDemoForm.tbDateTimeClientPropertiesChange(Sender: TObject);
begin
  (rcDateTimeClient.ClientProperties as TdxRangeControlDateTimeClientProperties).ScaleInterval := tbDateTimeClient.Position;
end;

procedure TdxRangeControlDemoForm.tbNumericClientPropertiesChange(
  Sender: TObject);
begin
  (rcNumericClient.ClientProperties as TdxRangeControlNumericClientProperties).ScaleInterval := tbNumericClient.Position;
end;

procedure TdxRangeControlDemoForm.FormCreate(Sender: TObject);
begin
  inherited;
  FNumericClientData := TList<TChartData>.Create;
  FDateTimeClientData2 := TList<TChartData>.Create;
  FDateTimeClientData1 := TList<TChartData>.Create;
  FDateTimeHeaderClientData1 := TDictionary<TDateTime, Integer>.Create;
  FDateTimeHeaderClientData2 := TDictionary<TDateTime, Integer>.Create;
  rcNumericClient.SelectedRangeMaxValue := 10;
  rcDateTimeClient.ClientProperties.MinValue := Date - 5;
  rcDateTimeClient.ClientProperties.MaxValue := Date + 5;
  rcDateTimeClient.SelectedRangeMinValue := Date - 1;
  rcDateTimeClient.SelectedRangeMaxValue := Date + 1;
  rcDateTimeHeaderClient.ClientProperties.MinValue := Date - 5;
  rcDateTimeHeaderClient.ClientProperties.MaxValue := Date + 5;
  rcDateTimeHeaderClient.SelectedRangeMinValue := Date;
  rcDateTimeHeaderClient.SelectedRangeMaxValue := Date + 2;
  InitializeChartData;
  Width := 850;
  Height := 650;
  Constraints.MinWidth := Width;
  Constraints.MinHeight := Height;
end;

procedure TdxRangeControlDemoForm.FormDestroy(Sender: TObject);
begin
  FreeAndNil(FDateTimeHeaderClientData2);
  FreeAndNil(FDateTimeHeaderClientData1);
  FreeAndNil(FDateTimeClientData1);
  FreeAndNil(FDateTimeClientData2);
  FreeAndNil(FNumericClientData);
  inherited;
end;

procedure TdxRangeControlDemoForm.FormResize(Sender: TObject);
const
  ControlCount = 3;
  Interval = 10;
var
  R: TRect;
  AHeight: Integer;
begin
  R := cxGroupBox5.ClientBounds;
  AHeight := (R.Height - 4 * 10) div 3;
  cxGroupBox1.Top := R.Top + 10;
  cxGroupBox1.Height := AHeight;
  cxGroupBox2.Top := cxGroupBox1.Top + AHeight + 10;
  cxGroupBox2.Height := AHeight;
  cxGroupBox3.Top := cxGroupBox2.Top + AHeight + 10;
  cxGroupBox3.Height := AHeight;
end;

procedure TdxRangeControlDemoForm.rcDateTimeClientDrawContent(ASender: TdxCustomRangeControl;
  ACanvas: TcxCanvas; AViewInfo: TdxRangeControlCustomClientViewInfo; var AHandled: Boolean);

  procedure DrawData(ADataSource: TList<TChartData>; const R: TRect; AColor, ABrushColor: TdxAlphaColor);
  var
    AData: TChartData;
    APos, AValue: Integer;
    APolygon: TPoints;
  begin
    dxGPPaintCanvas.BeginPaint(ACanvas.Handle, AViewInfo.Bounds);
    dxGPPaintCanvas.SmoothingMode := smAntiAlias;
    try
      if not IsDateTimeClientContentLineMode then
        AddPolygonPoint(APolygon, R.Left, R.Bottom);
      for AData in ADataSource do
        if InRange(AData.X, ASender.VisibleRangeMinValue - 1, ASender.VisibleRangeMaxValue + 1) then
        begin
          APos := ASender.GetPositionFromValue(AData.X);
          AValue := R.Bottom - AData.Y * R.Height div 100;
          AddPolygonPoint(APolygon, APos, AValue);
        end;
      if IsDateTimeClientContentLineMode then
        dxGPPaintCanvas.Polyline(APolygon, AColor)
      else
      begin
        AddPolygonPoint(APolygon, R.Right, R.Bottom);
        dxGPPaintCanvas.Polygon(APolygon, AColor, ABrushColor);
      end;
    finally
      dxGPPaintCanvas.EndPaint;
    end;
  end;

var
  R: TRect;
begin
  R := AViewInfo.Content.Bounds;
  DrawData(FDateTimeClientData2, R, $FFEE8C4B, $70EE8C4B);
  DrawData(FDateTimeClientData1, R, $FF6AA4D9, $906AA4D9);
end;

procedure TdxRangeControlDemoForm.rcDateTimeHeaderClientDrawContent(
  ASender: TdxCustomRangeControl; ACanvas: TcxCanvas; AViewInfo: TdxRangeControlCustomClientViewInfo; var AHandled: Boolean);
var
  AContentElements: TList<TdxRangeControlDateTimeHeaderClientContentElementViewInfo>;
  AElement: TdxRangeControlDateTimeHeaderClientContentElementViewInfo;
  I: Integer;
  AValue: Integer;
  ARect: TRect;
begin
  AContentElements := (AViewInfo.Content as TdxRangeControlDateTimeHeaderClientContentViewInfo).Elements;
  dxGPPaintCanvas.BeginPaint(ACanvas.Handle, AViewInfo.Bounds);
  dxGPPaintCanvas.SmoothingMode := smAntiAlias;
  try
    for I := 0 to AContentElements.Count - 1 do
    begin
      AElement := AContentElements[I];
      dxGPPaintCanvas.SaveClipRegion;
      try
        dxGPPaintCanvas.SetClipRect(AElement.Bounds, gmIntersect);
        if FDateTimeHeaderClientData1.TryGetValue(AElement.MinDate, AValue) then
        begin
          ARect := AElement.Bounds;
          ARect.Right := cxRectCenter(ARect).X - 2;
          ARect.Left := ARect.Right - 15;
          Inc(ARect.Bottom);
          ARect.Top := ARect.Bottom - AValue * ARect.Height div 100;
          dxGPPaintCanvas.Rectangle(ARect, $FF6AA4D9, $646AA4D9);
        end;
        if FDateTimeHeaderClientData2.TryGetValue(AElement.MinDate, AValue) then
        begin
          ARect := AElement.Bounds;
          ARect.Left := cxRectCenter(ARect).X + 2;
          ARect.Right := ARect.Left + 15;
          Inc(ARect.Bottom);
          ARect.Top := ARect.Bottom - AValue * ARect.Height div 100;
          dxGPPaintCanvas.Rectangle(ARect, $FFEE8C4B, $64EE8C4B);
        end;
      finally
        dxGPPaintCanvas.RestoreClipRegion;
      end;
    end;
  finally
    dxGPPaintCanvas.EndPaint;
  end;
end;

procedure TdxRangeControlDemoForm.InitializeChartData;
var
  ADateTime: TDateTime;
  AData: TChartData;
  I: Integer;
begin
  Randomize;
  ADateTime := rcDateTimeClient.ClientProperties.MinValue;
  while ADateTime <= rcDateTimeClient.ClientProperties.MaxValue do
  begin
    AData.X := ADateTime;
    AData.Y := RandomRange(2, 50);
    FDateTimeClientData1.Add(AData);
    AData.Y := RandomRange(52, 100);
    FDateTimeClientData2.Add(AData);
    ADateTime := IncHour(ADateTime, 3);
  end;
  for I := rcNumericClient.ClientProperties.MinValue to rcNumericClient.ClientProperties.MaxValue do
  begin
    AData.X := I;
    AData.Y := RandomRange(2, 100);
    FNumericClientData.Add(AData);
  end;
  ADateTime := rcDateTimeHeaderClient.ClientProperties.MinValue;
  while ADateTime <= rcDateTimeHeaderClient.ClientProperties.MaxValue do
  begin
    FDateTimeHeaderClientData1.Add(ADateTime, RandomRange(10, 90));
    FDateTimeHeaderClientData2.Add(ADateTime, RandomRange(10, 90));
    ADateTime := IncDay(ADateTime);
  end;
end;

function TdxRangeControlDemoForm.IsDateTimeClientContentLineMode: Boolean;
begin
  Result := rgDateTimeClientContentType.ItemIndex = 0;
end;

function TdxRangeControlDemoForm.IsNumberClientContentLineMode: Boolean;
begin
  Result := rgNumberClientContentType.ItemIndex = 0;
end;

procedure TdxRangeControlDemoForm.ShowRuler1Click(Sender: TObject);
var
  AIsChecked: Boolean;
begin
  AIsChecked := dxDemoIsMenuChecked(Sender);
  rcNumericClient.ShowRuler := AIsChecked;
  rcDateTimeClient.ShowRuler := AIsChecked;
  rcDateTimeHeaderClient.ShowRuler := AIsChecked;
end;

procedure TdxRangeControlDemoForm.ShowZoomscrollbar1Click(Sender: TObject);
var
  AIsChecked: Boolean;
begin
  AIsChecked := dxDemoIsMenuChecked(Sender);
  rcNumericClient.ShowZoomScrollBar := AIsChecked;
  rcDateTimeClient.ShowZoomScrollBar := AIsChecked;
  rcDateTimeHeaderClient.ShowZoomScrollBar := AIsChecked;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressEditors Library/CBuilder/RangeControlDemo/RangeControlDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "RangeControlDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "BaseForm"
#pragma link "cxCheckBox"
#pragma link "cxCheckComboBox"
#pragma link "cxContainer"
#pragma link "cxControls"
#pragma link "cxDropDownEdit"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxGroupBox"
#pragma link "cxLabel"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxMaskEdit"
#pragma link "cxRadioGroup"
#pragma link "cxTextEdit"
#pragma link "cxTrackBar"
#pragma link "dxRangeControl"
#pragma resource "*.dfm"
TdxRangeControlDemoForm *dxRangeControlDemoForm;
//---------------------------------------------------------------------------
__fastcall TdxRangeControlDemoForm::TdxRangeControlDemoForm(TComponent* Owner)
	: TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TdxRangeControlDemoForm::rgNumberClientContentTypePropertiesEditValueChanged(TObject *Sender)

{
  rcNumericClient->ContentChanged();
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::tbNumericClientPropertiesChange(TObject *Sender)

{
  dynamic_cast<TdxRangeControlNumericClientProperties*>(rcNumericClient->ClientProperties)->ScaleInterval = tbNumericClient->Position;
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::rgDateTimeClientContentTypePropertiesEditValueChanged(TObject *Sender)
{
  rcDateTimeClient->ContentChanged();
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::tbDateTimeClientPropertiesChange(TObject *Sender)
{
  dynamic_cast<TdxRangeControlDateTimeClientProperties*>(rcDateTimeClient->ClientProperties)->ScaleInterval = tbDateTimeClient->Position;
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::chbAutoFormatScaleCaptionsPropertiesEditValueChanged(TObject *Sender)
{
  dynamic_cast<TdxRangeControlDateTimeHeaderClientProperties*>(rcDateTimeHeaderClient->ClientProperties)->AutoFormatScaleCaptions = chbAutoFormatScaleCaptions->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::cxCheckComboBox1PropertiesEditValueChanged(TObject *Sender)
{
  TdxRangeControlDateTimeHeaderClientProperties *AProperties = dynamic_cast<TdxRangeControlDateTimeHeaderClientProperties*>(rcDateTimeHeaderClient->ClientProperties);
  TdxRangeControlDateTimeScales *AScales = AProperties->Scales;
  for (int i = 0; i < cxCheckComboBox1->Properties->Items->Count; i++)
    AScales->GetScale(TdxRangeControlDateTimeScaleUnit(i + (rcduDay)))->Visible = (cxCheckComboBox1->States[i] == cbsChecked);
}
//---------------------------------------------------------------------------
void __fastcall TdxRangeControlDemoForm::InitializeChartData()
{
  Randomize();
  TDateTime ADateTime = rcDateTimeClient->ClientProperties->MinValue;
  TdxRangeControlDateTimeClientHelper *A = new TdxRangeControlDateTimeClientHelper();
  while (ADateTime <= rcDateTimeClient->ClientProperties->MaxValue)
  {
	FDateTimeClientData1.Length = FDateTimeClientData1.Length + 1;
	FDateTimeClientData1[FDateTimeClientData1.High].X = ADateTime;
	FDateTimeClientData1[FDateTimeClientData1.High].Y = RandomRange(2, 50);
	FDateTimeClientData2.Length = FDateTimeClientData2.Length + 1;
	FDateTimeClientData2[FDateTimeClientData2.High].X = ADateTime;
	FDateTimeClientData2[FDateTimeClientData2.High].Y = RandomRange(52, 100);

	ADateTime = A->IncDate(ADateTime, rcduHour, 3);
  };
  for (int i = rcNumericClient->ClientProperties->MinValue; i <= rcNumericClient->ClientProperties->MaxValue; i++)
  {
	FNumericClientData.Length = FNumericClientData.Length + 1;
	FNumericClientData[FNumericClientData.High].X = i;
	FNumericClientData[FNumericClientData.High].Y = RandomRange(2, 100);
  };
  TDateTime ADateTime1 = rcDateTimeHeaderClient->ClientProperties->MinValue;
  while (ADateTime1 <= rcDateTimeHeaderClient->ClientProperties->MaxValue)
  {
	FDateTimeHeaderClientData1.Length = FDateTimeHeaderClientData1.Length + 1;
	FDateTimeHeaderClientData1[FDateTimeHeaderClientData1.High].X = ADateTime1;
	FDateTimeHeaderClientData1[FDateTimeHeaderClientData1.High].Y = RandomRange(10, 90);
	FDateTimeHeaderClientData2.Length = FDateTimeHeaderClientData2.Length + 1;
	FDateTimeHeaderClientData2[FDateTimeHeaderClientData2.High].X = ADateTime1;
	FDateTimeHeaderClientData2[FDateTimeHeaderClientData2.High].Y = RandomRange(10, 90);

	ADateTime1 = A->IncDate(ADateTime1, rcduDay);
  };
  delete A;
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::FormCreate(TObject *Sender)
{
  rcNumericClient->SelectedRangeMaxValue = 10;
  rcDateTimeClient->ClientProperties->MinValue = Date() - 5;
  rcDateTimeClient->ClientProperties->MaxValue = Date() + 5;
  rcDateTimeClient->SelectedRangeMinValue = Date() - 1;
  rcDateTimeClient->SelectedRangeMaxValue = Date() + 1;
  rcDateTimeHeaderClient->ClientProperties->MinValue = Date() - 5;
  rcDateTimeHeaderClient->ClientProperties->MaxValue = Date() + 5;
  rcDateTimeHeaderClient->SelectedRangeMinValue = Date();
  rcDateTimeHeaderClient->SelectedRangeMaxValue = Date() + 2;
  InitializeChartData();
  Width = 850;
  Height = 650;
  Constraints->MinWidth = Width;
  Constraints->MinHeight = Height;
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::Animation1Click(TObject *Sender)
{
  bool AChecked = dynamic_cast<TMenuItem*>(Sender)->Checked;
  rcNumericClient->Animation = AChecked;
  rcDateTimeClient->Animation = AChecked;
  rcDateTimeHeaderClient->Animation = AChecked;
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::ShowRuler1Click(TObject *Sender)
{
  bool AChecked = dynamic_cast<TMenuItem*>(Sender)->Checked;
  rcNumericClient->ShowRuler = AChecked;
  rcDateTimeClient->ShowRuler = AChecked;
  rcDateTimeHeaderClient->ShowRuler = AChecked;
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::ShowZoomscrollbar1Click(TObject *Sender)

{
  bool AChecked = dynamic_cast<TMenuItem*>(Sender)->Checked;
  rcNumericClient->ShowZoomScrollBar = AChecked;
  rcDateTimeClient->ShowZoomScrollBar = AChecked;
  rcDateTimeHeaderClient->ShowZoomScrollBar = AChecked;
}
//---------------------------------------------------------------------------
bool __fastcall TdxRangeControlDemoForm::IsNumberClientContentLineMode()
{
  return !rgNumberClientContentType->ItemIndex;
}
//---------------------------------------------------------------------------
bool __fastcall TdxRangeControlDemoForm::IsDateTimeClientContentLineMode()
{
  return !rgDateTimeClientContentType->ItemIndex;
}
//---------------------------------------------------------------------------
void __fastcall TdxRangeControlDemoForm::AddPolygonPoint(TPoints &APolygon, int X, int Y)
{
  APolygon.Length = APolygon.Length + 1;
  APolygon[APolygon.High].x = X;
  APolygon[APolygon.High].y = Y;
}
//---------------------------------------------------------------------------
void __fastcall TdxRangeControlDemoForm::GetPoints(TdxRangeControl *ARangeControl, TdxRangeControlCustomClientViewInfo *AViewInfo,
  TRect R, TChartDatas ADataSource, TPoints &APolygon)
{
  int AVisibleRangeMinValue = ARangeControl->VisibleRangeMinValue;
  int AVisibleRangeMaxValue = ARangeControl->VisibleRangeMaxValue;
  for (int i = 0; i < ADataSource.Length; i++)
	{
	if (InRange(ADataSource[i].X, AVisibleRangeMinValue - 1, AVisibleRangeMaxValue + 1))
	{
	  int APos = ARangeControl->GetPositionFromValue(ADataSource[i].X);
	  int AValue = R.Bottom - ADataSource[i].Y * R.Height() / 100;
      AddPolygonPoint(APolygon, APos, AValue);
	};
   }
}
//---------------------------------------------------------------------------
void __fastcall TdxRangeControlDemoForm::DrawNumericData(
  TcxCanvas *ACanvas, TdxRangeControlCustomClientViewInfo *AViewInfo, TPoints APolygon, TRect R, TdxAlphaColor AColor)
{
	dxGPPaintCanvas()->BeginPaint(ACanvas->Handle, AViewInfo->Bounds);
	dxGPPaintCanvas()->SmoothingMode = smAntiAlias;
	try
	{
	  if (IsNumberClientContentLineMode())
		dxGPPaintCanvas()->Polyline(&APolygon[0], APolygon.Length - 1, AColor, 1, psSolid);
	  else
		{
		  for (int i = 0; i < APolygon.Length; i++)
		  {
			int APos = R.Bottom;
			dxGPPaintCanvas()->Line(APolygon[i].x, APos, APolygon[i].x, APolygon[i].y, AColor, 3);
		  }
		}
	}
	__finally
	  {
		dxGPPaintCanvas()->EndPaint();
	  };
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::rcNumericClientDrawContent(TdxCustomRangeControl *ASender,
	TcxCanvas *ACanvas, TdxRangeControlCustomClientViewInfo *AViewInfo,
	bool &AHandled)
{
	TPoints APolygon;
	APolygon.Length = 0;
	GetPoints(rcNumericClient, AViewInfo, AViewInfo->Content->Bounds, FNumericClientData, APolygon);
	DrawNumericData(ACanvas, AViewInfo, APolygon, AViewInfo->Content->Bounds, 0xFFEE8C4B);
}
//---------------------------------------------------------------------------
void __fastcall TdxRangeControlDemoForm::DrawDateTimeData(
	TcxCanvas *ACanvas, TdxRangeControlCustomClientViewInfo *AViewInfo, TPoints APolygon,
	TRect R, TdxAlphaColor AColor, TdxAlphaColor ABrushColor)
{
	dxGPPaintCanvas()->BeginPaint(ACanvas->Handle, AViewInfo->Bounds);
	dxGPPaintCanvas()->SmoothingMode = smAntiAlias;
	try
	{
	  if (IsDateTimeClientContentLineMode())
		dxGPPaintCanvas()->Polyline(&APolygon[0], APolygon.Length - 1, AColor, 1, psSolid);
	  else
        dxGPPaintCanvas()->Polygon(&APolygon[0], APolygon.Length - 1, AColor, ABrushColor, 1, psSolid);
	}
	__finally
	  {
		dxGPPaintCanvas()->EndPaint();
	  };
}
//---------------------------------------------------------------------------
void __fastcall TdxRangeControlDemoForm::rcDateTimeClientDrawContent(TdxCustomRangeControl *ASender,
	TcxCanvas *ACanvas, TdxRangeControlCustomClientViewInfo *AViewInfo,
	bool &AHandled)
{
	TPoints APolygon;
	TRect R = AViewInfo->Content->Bounds;
	APolygon.Length = 0;
	if (!IsDateTimeClientContentLineMode())
	  AddPolygonPoint(APolygon, R.Left, R.Bottom);
	GetPoints(rcDateTimeClient, AViewInfo, R, FDateTimeClientData2, APolygon);
	if (!IsDateTimeClientContentLineMode())
	  AddPolygonPoint(APolygon, R.Right, R.Bottom);
	DrawDateTimeData(ACanvas, AViewInfo, APolygon, R, 0xFFEE8C4B, 0x70EE8C4B);
	APolygon.Length = 0;
	if (!IsDateTimeClientContentLineMode())
	  AddPolygonPoint(APolygon, R.Left, R.Bottom);
	GetPoints(rcDateTimeClient, AViewInfo, R, FDateTimeClientData1, APolygon);
	if (!IsDateTimeClientContentLineMode())
	  AddPolygonPoint(APolygon, R.Right, R.Bottom);
	DrawDateTimeData(ACanvas, AViewInfo, APolygon, R, 0xFF6AA4D9, 0x906AA4D9);
}
//---------------------------------------------------------------------------
int __fastcall TdxRangeControlDemoForm::GetDateTimeHeaderValue(TChartDatas ADataSource, TdxRangeControlDateTimeHeaderClientContentElementViewInfo *AElement)
{
	for (int i = 0; i < ADataSource.Length; i++) {
	  if ((ADataSource[i].X >= AElement->MinDate)&&(ADataSource[i].X < AElement->MaxDate)) {
		return(ADataSource[i].Y);
	  }
	}
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::rcDateTimeHeaderClientDrawContent(TdxCustomRangeControl *ASender,
          TcxCanvas *ACanvas, TdxRangeControlCustomClientViewInfo *AViewInfo,
          bool &AHandled)
{
    TdxRangeControlDateTimeHeaderClientContentViewInfo *AContentElementsViewInfo = dynamic_cast<TdxRangeControlDateTimeHeaderClientContentViewInfo*>(AViewInfo->Content);
	dxGPPaintCanvas()->BeginPaint(ACanvas->Handle, AViewInfo->Bounds);
	dxGPPaintCanvas()->SmoothingMode = smAntiAlias;
	try
	{
		for (int i = 0; i < AContentElementsViewInfo->Elements->Count; i++)
		{
			TdxRangeControlDateTimeHeaderClientContentElementViewInfo *AElement = AContentElementsViewInfo->Elements->Items[i];
			dxGPPaintCanvas()->SaveClipRegion();
			try
			{
				dxGPPaintCanvas()->SetClipRect(AElement->Bounds, gmIntersect);

				int AValue = GetDateTimeHeaderValue(FDateTimeHeaderClientData1, AElement);
				TRect ARect = AElement->Bounds;
				ARect.Right = cxRectCenter(ARect).x - 2;
				ARect.Left = ARect.Right - 15;
				ARect.Bottom ++;
				ARect.Top = ARect.Bottom - AValue * ARect.Height() / 100;
				dxGPPaintCanvas()->Rectangle(ARect, 0xFF6AA4D9, 0x646AA4D9);

                AValue = GetDateTimeHeaderValue(FDateTimeHeaderClientData2, AElement);
				ARect = AElement->Bounds;
				ARect.Left = cxRectCenter(ARect).x + 2;
				ARect.Right = ARect.Left + 15;
				ARect.Bottom ++;
				ARect.Top = ARect.Bottom - AValue * ARect.Height() / 100;
				dxGPPaintCanvas()->Rectangle(ARect, 0xFFEE8C4B, 0x64EE8C4B);
			}
			__finally
			{
				dxGPPaintCanvas()->RestoreClipRegion();
			};
		 }
	  }
	__finally
	  {
		dxGPPaintCanvas()->EndPaint();
	  }
}
//---------------------------------------------------------------------------


void __fastcall TdxRangeControlDemoForm::FormResize(TObject *Sender)
{
  int ControlCount = 3;
  int Interval = 10;
  TRect R = cxGroupBox5->ClientBounds;
  int AHeight = (R.Height() - 4 * Interval) / ControlCount;
  cxGroupBox1->Top = R.Top + Interval;
  cxGroupBox1->Height = AHeight;
  cxGroupBox2->Top = cxGroupBox1->Top + AHeight + Interval;
  cxGroupBox2->Height = AHeight;
  cxGroupBox3->Top = cxGroupBox2->Top + AHeight + Interval;
  cxGroupBox3->Height = AHeight;
}
//---------------------------------------------------------------------------

void __fastcall TdxRangeControlDemoForm::FormDestroy(TObject *Sender)
{
//
}
//---------------------------------------------------------------------------

